### PR TITLE
Tempus: Build Fails with ETI=OFF

### LINKED ISSUE #8548: Tempus:  Build Fails with ETI=OFF
## Bug Report
@trilinos/tempus 

### Description
When ETI is OFF ( -D Tempus_ENABLE_EXPLICIT_INSTANTIATION:BOOL=OFF ), Tempus fails to build.

### Steps to Reproduce
1. SHA1:  90422f3c55f
1. Configure script:  [build_tempus_s1037643_mac_gcc.sh.txt](https://github.com/trilinos/Trilinos/files/5784204/build_tempus_s1037643_mac_gcc.sh.txt)
1. Configure log:  [Tempus_config.log](https://github.com/trilinos/Trilinos/files/5784213/Tempus_config.log)
1. Build log:  [Tempus_ETI_OFF.log](https://github.com/trilinos/Trilinos/files/5784196/Tempus_ETI_OFF.log)

### PR DESCRIPTION
**Pull request by ccober6**
Removed some circular includes that Roger pointed out.
 * In BDF2 and BackwardEuler where startup/predictor
   Stepper are set by string.  Moved to set the
   startup/predictor Stepper by stepper, which does
   not require the StepperFactory (and therefore the
   circular includes).
 * Also removed some circular includes in the AppAction
   which should not be there.  Left the forward declarations.

Passes all tests, but still fails with ETI=OFF.

@trilinos/tempus 

## Motivation
Fix in order to build with ETI=OFF.

## Related Issues

* Closes #8548

### CODE REVIEW

**Comment by ccober6** [comment]
Hi @rppawlo, 

Here are the changes I have made so far.  As noted above, I removed the circular includes I could find.  But I am still seeing the errors related to StepperFactory.  For some reason it believes that 
 * the StepperFactory is a variable template at Tempus_StepperFactory_impl.hpp:17 instead of a scoping for the createStepper() function, and
 * the StepperFactory class definition at Tempus_StepperFactory_decl.hpp:37 is a redefinition of 'StepperFactory' at Tempus_StepperFactory_impl.hpp:17

It appears that these are in inverted order?

**Review by ikalash — APPROVED** [review]
Looks good.  This should not cause problems to Albany after I merge in PRs that go hand-in-hand with this PR on the Albany side.

**Comment by ccober6** [comment]
Thanks @ikalash, but I am still working on this.  The StepperFactory and Steppers have a bad circular dependence that I am trying to clear up.  :(

**Comment by ikalash** [comment]
Oh...  sorry I didn't realize that.  It looked like the autotester needed another approval to get re-kicked off.

**Comment by ccober6** [comment]
@rppawlo I think this is ready for review.  I might need to make some changes once #8372 goes thru.  Thanks for the help!

**Comment by ccober6** [comment]
@rppawlo I think this is ready for merge.  Would you have time to review it?

**Review by rppawlo — DISMISSED** [review]
Looks good to me! A couple of suggestions:

1. With all extra file includes, you might just want to test build times to make sure they don't increase too much.
2. The nonmember ctors could be converted to support ETI as well. Build times might be impacted by this - I think the initial implementation did ETI for these since they were nested in the factory object.

**Comment by ccober6** [comment]
@ikalash @michael-a-hansen @sconde Can you check this PR?

The major change for the user should be the use of createStepper*() functions.  To eliminate a circular include issue, these functions were moved to the individual Steppers and made as nonmember constructors.  Thus Stepper construction that used StepperFactory, e.g.,
```
  auto sf = Teuchos::rcp(new Tempus::StepperFactory<double>());
  auto stepperFE = sf->createStepperForwardEuler(model, Teuchos::null);
```
and returned a StepperForwardEuler object, can be replaced with the nonmember constructor, e.g.,
```
  auto stepperFE = Tempus::createStepperForwardEuler(modelME, Teuchos::null);
```
which also returns a StepperForwardEuler object.

You can still use the StepperFactory to create various steppers, but it will return the base class Stepper (which could be dynamical cast to a StepperForwardEuler if you like).

I also cleaned up some includes, which is one reason for the number of files touched.

Let me know if you have any questions.

**Comment by ikalash** [comment]
@ccober6 : can you please merge develop into the branch and let me know when you do?  I'm getting an error from Kokkos/Tpetra in my build, which was recently resolved in develop.

```
     |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /nightlyAlbanyTests/Results/Trilinos/packages/tpetra/core/src/Tpetra_Details_FixedHashTable_Serial.cpp:49:
/nightlyAlbanyTests/Results/Trilinos/packages/tpetra/core/src/Tpetra_Details_FixedHashTable_def.hpp: In instantiation of ‘void Tpetra::Details::FixedHashTable<KeyType, ValueType, DeviceType>::init(const keys_type&, ValueType, KeyType, KeyType, KeyType, KeyType, bool) [with KeyType = int; ValueType = int; DeviceType = Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>; Tpetra::Details::FixedHashTable<KeyType, ValueType, DeviceType>::keys_type = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace> >]’:
/nightlyAlbanyTests/Results/Trilinos/packages/tpetra/core/src/Tpetra_Details_FixedHashTable_Serial.cpp:72:3:   required from here
/nightlyAlbanyTests/Results/Trilinos/packages/tpetra/core/src/Tpetra_Details_FixedHashTable_def.hpp:1097:17: error: redeclaration of ‘using ViewAllocateWithoutInitializing = struct Kokkos::Impl::ViewCtorProp<Kokkos::Impl::WithoutInitializing_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Kokkos::Impl::ViewAllocateWithoutInitializingBackwardCompat>’
 1097 |   using Kokkos::ViewAllocateWithoutInitializing;
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Also, we don't typically use Trilinos with ETI = OFF anymore, so I will test with our usual use case (ETI=ON).

**Comment by ccober6** [comment]
@ikalash Will do.  The hazards of rebasing to soon.  :)

Yeah, same here on testing ETI=OFF, but there is a plan to have a Trilinos build that will test this and prevent regression.  So I did an early build which pointed out that we had a circular set of includes that needed to be fixed in order to build with ETI=OFF.  I am not sure on the timeline for the build but when you get a chance you may want to test it also for Piro?

**Comment by ikalash** [comment]
I will test it as soon as I can get it to compiler.  It appears there are unresolved conflicts now stemming from a merge (after I pulled the branch): 

```
[ 93%] Building CXX object packages/tempus/src/CMakeFiles/tempus.dir/Tempus_StepperImplicit.cpp.o
In file included from /nightlyAlbanyTests/Results/Trilinos/build/packages/tempus/src/Tempus_StepperIMEX_RK.hpp:1,
                 from /nightlyAlbanyTests/Results/Trilinos/packages/tempus/src/Tempus_StepperIMEX_RK.cpp:12:
/nightlyAlbanyTests/Results/Trilinos/packages/tempus/src/Tempus_StepperIMEX_RK_decl.hpp:488:1: error: version control conflict marker in file
  488 | <<<<<<< HEAD
      | ^~~~~~~
/nightlyAlbanyTests/Results/Trilinos/packages/tempus/src/Tempus_StepperIMEX_RK_decl.hpp:489:40: error: expected unqualified-id before ‘pl’
  489 |   Teuchos::RCP<Teuchos::ParameterList> pl)
      |                                        ^~
/nightlyAlbanyTests/Results/Trilinos/packages/tempus/src/Tempus_StepperIMEX_RK_decl.hpp:489:39: error: expected ‘)’ before ‘pl’
  489 |   Teuchos::RCP<Teuchos::ParameterList> pl)
      |                                       ^~~
      |                                       )
/nightlyAlbanyTests/Results/Trilinos/packages/tempus/src/Tempus_StepperIMEX_RK_decl.hpp:485:21: note: to match this ‘(’
  485 | createStepperIMEX_RK(
      |                     ^
/nightlyAlbanyTests/Results/Trilinos/packages/tempus/src/Tempus_StepperIMEX_RK_decl.hpp:489:40: error: expected initializer before ‘pl’
  489 |   Teuchos::RCP<Teuchos::ParameterList> pl)
      |                                        ^~
/nightlyAlbanyTests/Results/Trilinos/packages/tempus/src/Tempus_StepperIMEX_RK_decl.hpp:503:1: error: version control conflict marker in file
  503 | =======
      | ^~~~~~~
```

**Comment by ccober6** [comment]
Hi @ikalash.  So I think the branch is ok, as the autotester is passing (see above).  My guess on what is happening is I did a rebase/squash with a force push on this branch (suggested practice to keep the revision history clean/simple with only passing commits).  This can mess-up other developer's branch, because it has changed the branch's history.  I have fixed this in the past with something like
```
git fetch origin
git reset --hard origin/8548-Tempus-Build-Fails-with-ETI-OFF
```
In fact, I needed to do this on my cee-lan repo when testing there for this branch (i.e., had the branch checked out already when I did the rebase/squash).  Let me know if this does not help.

**Comment by ikalash** [comment]
@ccober6 : I just tested this PR and everything is good for Albany and Albany LCM.

**Comment by ccober6** [comment]
@sconde A couple things.  Have you had a chance to test this against Drekar?  And if you have a "minute"  Can you review this?

**Comment by sconde** [comment]
> @sconde A couple things. Have you had a chance to test this against Drekar? And if you have a "minute" Can you review this?

My apologies. Will work on that now

File: packages/tempus/src/Tempus_IntegratorBasic_impl.hpp [hunk 1]
```diff
@@ -97,16 +95,12 @@ void IntegratorBasic<Scalar>::setStepper(
   using Teuchos::RCP;
   using Teuchos::ParameterList;
 
-  if (stepper_ == Teuchos::null) {
```
Comment (sconde) [inline packages/tempus/src/Tempus_IntegratorBasic_impl.hpp:100]: Is there a reason for eliminating the guard if the stepper is non-null?

**Comment by sconde** [comment]
@ccober6 build is failing with Drekar, so I'm working on that at the moment. 

I see you've removed some comments and I would suggest adding/updating doxygen to the files modified.

File: packages/tempus/src/Tempus_IntegratorBasic_impl.hpp (hunk 1 above)
Comment (ccober6) [reply packages/tempus/src/Tempus_IntegratorBasic_impl.hpp:100]: Maybe we should leave it as is.  I thought there was some wrong logic here, but it is probably ok.  I am planning dropping this function anyway, as a function to set the Stepper from models as the argument is confusing.  If the function is setStepper(), it should set the stepper, not set the models on the stepper. 

I will return to the original coding.

Also I am not liking how we set the steppers for StepperOperatorSplit steppers, i.e., give a list of steppers and a list of models and make them match.  I think we should remove this functionality and exclusively go with completely setup a stepper with its model, and then add it to the StepperOperatorSpit stepper via addStepper().  That way it is clear what models go with what stepper.  thoughts?

**Comment by ccober6** [comment]
@rppawlo  I made the "Converted Nonmember Ctor to ETI" changes, and rebased with develop for @ikalash .  Can you reapprove this for merge?  Thanks!

> @ccober6 build is failing with Drekar, so I'm working on that at the moment.
> 
> I see you've removed some comments and I would suggest adding/updating doxygen to the files modified.

@sconde can you be specific which comments you noted being removed?

**Comment by ccober6** [comment]
Not sure what happened on the previous comment, but ...

>     I see you've removed some comments and I would suggest adding/updating doxygen to the files modified.

@sconde can you be specific which comments you noted being removed?  Are you talking about the comments that labeled the sections of onclude files?  If so I think it is pretty apparent that the include files are from the various packages (e.g., Thyra, Teuchos, Tempus, ...) and we do not need comments to state that.

File: packages/tempus/src/Tempus_IntegratorBasic_impl.hpp (hunk 1 above)
Comment (ccober6) [reply packages/tempus/src/Tempus_IntegratorBasic_impl.hpp:100]: Ah, yes.  Stepper::createSubSteppers() with a single argument was a no-op function!  So the function did nothing when stepper_ != null.  The whole thing is ugly and not tested.  Thus I think we just need to remove it in a future PR.

Comment (ccober6) [reply packages/tempus/src/Tempus_IntegratorBasic_impl.hpp:100]: Let's not worry about this in this PR.  I would suggest approving this PR and make fixes in another PR.

**Review by rppawlo — APPROVED** [review]

**Comment by ccober6** [comment]
@sconde Is Drekar ready for this merge?

**Review by sconde — APPROVED** [review]

**Comment by sconde** [comment]
@ccober6 yes, it is. But do you think you can merge https://github.com/trilinos/Trilinos/pull/8734 into this PR's branch? Or should it be done separaterly?

**Comment by ccober6** [comment]
@sconde Let's go ahead and do them separately as I do not think there will be any conflicts.  If they are, it should be fairly straight forward to resolve.  :)

**Pull request merged**